File: 2024/Leetcode/problem20.py
class Solution:
    def isValid(self, s: str) -> bool:
        myStack = []
        matches = {'}': '{', ']': '[', ')': '('}

        for char in s:
            if char in matches.values():
                myStack.append(char)
            elif len(myStack) != 0 and matches[char] == myStack[-1]:
                myStack.pop()
            else:
                return False
        return len(myStack) == 0

File: 2024/Leetcode/problem20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> myStack;
        unordered_map<char, char> matches = {{'(', ')'}, {'{', '}'}, {'[', ']'}};

        for (int i = 0; i < s.size(); i++) {
            if (matches.find(s[i]) != matches.end()) {
                myStack.push(s[i]);
            } else if (myStack.size() != 0 and matches[myStack.top()] == s[i]) {
                myStack.pop();
            } else {
                return false;
            }
        }
        return myStack.empty();
    }
};